common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -0,0 +1,966 @@
+          // the block to disk. If the block is still not written to disk after all #onData
+          // invocations, the final #onComplete invocation is the last attempt to write the
+          // block to disk. If we still couldn't write this block to disk after this, we give up
+          // on this block push request and respond failure to client. We could potentially
+          // buffer the block longer or wait for a few iterations inside #onData or #onComplete
+          // to increase the chance of writing the block to disk, however this would incur more
+          // memory footprint or decrease the server processing throughput for the shuffle
+          // service. In addition, during test we observed that by randomizing the order in
+          // which clients sends block push requests batches, only ~0.5% blocks failed to be
+          // written to disk due to this reason. We thus decide to optimize for server
+          // throughput and memory usage.
+          if (deferredBufs == null) {
+            deferredBufs = new LinkedList<>();
+          }
+          // Write the buffer to the in-memory deferred cache. Since buf is a slice of a larger
+          // byte buffer, we cache only the relevant bytes not the entire large buffer to save
+          // memory.
+          ByteBuffer deferredBuf = ByteBuffer.allocate(buf.remaining());
+          deferredBuf.put(buf);
+          deferredBuf.flip();
+          deferredBufs.add(deferredBuf);
+        }
+      }
+    }
+
+    @Override
+    public void onComplete(String streamId) throws IOException {
+      // TODO should the merge manager check for the merge completion ratio here and finalize
+      // TODO shuffle merge if appropriate? So the merge manager can potentially finalize early
+      // TODO and the file channel can be closed even if finalize merge request is somehow not
+      // TODO received from the driver? If so, then we need to know # maps for this shuffle.
+
+      synchronized (partitionInfo) {
+        logger.trace("{} shuffleId {} reduceId {} onComplete invoked", appShuffleId.appId,
+          appShuffleId.shuffleId, reduceId);
+        Map<Integer, AppShufflePartitionInfo> shufflePartitions =
+          mergeManager.partitions.get(appShuffleId);
+        // When this request initially got to the server, the shuffle merge finalize request
+        // was not received yet. By the time we finish reading this message, the shuffle merge
+        // however is already finalized. We should thus respond RpcFailure to the client.
+        if (shufflePartitions == null || !shufflePartitions.containsKey(reduceId)) {
+          deferredBufs = null;
+          throw new RuntimeException(String.format("Block %s %s", msg.streamId,
+            ErrorHandler.BlockPushErrorHandler.TOO_LATE_MESSAGE_SUFFIX));
+        }
+        // Check if we can commit this block
+        if (allowedToWrite()) {
+          isWriting = true;
```

Comment:
Do we really need this here in the `onComplete` method (when decision is only made based by `isWriting` value in the `onFailure` method)? 

What I mean is the`onComplete` method by definition is only called when all the data is received and everything went fine. And either [onComplete called](https://github.com/apache/spark/blob/0fa7d8051416db67f01edb8c93c9e2e2607d7b31/common/network-common/src/main/java/org/apache/spark/network/client/StreamInterceptor.java#L89)  or [`onFailure`](https://github.com/apache/spark/blob/0fa7d8051416db67f01edb8c93c9e2e2607d7b31/common/network-common/src/main/java/org/apache/spark/network/client/StreamInterceptor.java#L84).

Reply:
If there is an exception from `streamHandler.onComplete` [here](https://github.com/apache/spark/blob/56c623e98c54fdb4d47c9264ae1b282ecb2b7291/common/network-common/src/main/java/org/apache/spark/network/server/TransportRequestHandler.java#L211), then `streamHandler.onFailure` will also be called in the [catch block](https://github.com/apache/spark/blob/56c623e98c54fdb4d47c9264ae1b282ecb2b7291/common/network-common/src/main/java/org/apache/spark/network/server/TransportRequestHandler.java#L217).
This is why, we need `isWriting=true` in `onComplete` because there could be an IOException while writing to the file.

Reply:
Thanks, I missed that one!